### PR TITLE
Reintroduce spotlight/"important traits" feature

### LINKED ISSUE #73785: Reintroduce spotlight feature ("important traits")
I recently noticed that the "Important traits" feature (called "spotlight") was removed in https://github.com/rust-lang/rust/pull/69514

The stated justification was a twitter poll, however from going through the responses there it seemed like people hadn't _noticed_ it, not that people didn't find it useful. I don't think removing a feature is the right call if people aren't noticing it, that's a UX concern. It's also not that much code.

Can we reintroduce this UI as a little bubble on the return type? Putting it in the sidebar as a tiny circle was guaranteed to make it disappear. I found this feature quite useful, especially for iterators. In particular, my experience resonates with @Mark-Simulacrum's: https://github.com/rust-lang/rust/pull/69514#issuecomment-592159146 you need to click, scroll to the iterator impl, and then compare pages and substitute return types, whereas here it's all in one place.

I think we have a good indication that we should experiment with the UI and find something that works.

Ultimately, even if a feature is useful for some and not most, we shouldn't be removing it unless it's a net negative impact on the "most", which is definitely not clear here.

(I kinda wish that removing this feature had been FCPd)

cc @GuillaumeGomez @kinnison @rust-lang/rustdoc

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/rust-lang/rust/issues/73785

This PR reintroduces the "spotlight" ("important traits") feature.

A couple changes have been made:

As there were concerns about its visibility, it has been moved to be next to the return type, as opposed to being on the side.

It also no longer produces a modal, it shows the traits on hover, and it can be clicked on to pin the hover bubble.

image

image

It also works fine on mobile:

image

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Sorry but I'm still not in favor of this feature and the @rust-lang/rustdoc already voted to opt it out. It has been proven to be more noise than useful (and not only from the twitter thread linked in the issue).

**Comment by Mark-Simulacrum** [comment]
I feel like closing this is premature. I don't see an FCP for removal anywhere, and based on @kinnison's comments on https://github.com/rust-lang/rust/issues/73785 there's not really consensus amongst T-rustdoc that this isn't a good feature to have, rather, it was deemed "not horrible" to remove given your (strong?) opinion, especially because reverting that decision was relatively easy.

I for one continue to miss this feature pretty much every time I'm working with some non-obvious iterator where knowing the item type is super useful. The same goes for interactions with futures libraries, where it's particularly painful to figure out the item type.

If the problem is the choice of trait to display, IMO, we can require that the trait must be annotated with something like `#[rustdoc::notable_trait]` or whatever (fine to make unstable for now).

I don't think I've seen any discussion of *why* you're against it beyond it being too much information -- but the key point to me is that this is *just enough* information. I don't usually care about anything except for the Item/Output associated type, I already have a pretty good idea about the methods etc that Iterator and/or FutureExt or so provide.

**Comment by GuillaumeGomez** [comment]
Like I said, this decision doesn't come out of nowhere. It had a few downsides but the main one was simply that users didn't use it (and I don't have this information only from the twitter thread like I said). We kept this feature around for a while and whenever I talked about it, people just realized it existed and didn't find it useful in most cases and often they just didn't understand what it was used for.

The other big issue was the amount of JS code needed for it. Maintaining something used by a very small minority of users is not worth it considering how much time it takes me to maintain them. Also, "important traits" or whatever the name still remains a very anecdotal information.

Anyway, like I said, the rustdoc team opted it. If you want to debate about this feature, we can do it in an issue so no one loses more time than necessary writing code to get it closed like this (I really feel bad about that because of the time @Manishearth took to write this PR...). But as is, this is a no-go.

**Comment by GuillaumeGomez** [comment]
So after discussion with @Manishearth, we came to this conclusion: a few more steps will be required to reach a good enough level for everyone to be able to enjoy this feature, and in order to do so, we'll do it in multiple steps (asking for feedbacks along the road).

For the current PR, I propose the following changes:
* Please reword it to make it more clear.
* No popup, just a hover (eventually click) event handling for the text to appear (like a bubble, not a popup). The click handling is for the mobile devices.
* If possible, as little JS as possible (some CSS would allow us to handle the bubble easily on hover/click through a class).

**Comment by Manishearth** [comment]
@GuillaumeGomez made all the changes. The JS now used by this feature is minimal.

File: src/librustdoc/html/static/main.js [hunk 1]
```diff
@@ -2621,6 +2624,16 @@ function defocusSearchBar() {
         });
     }());
 
+    function showImportantTraits(content) {
+        let list = content.classList
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2628]: What is `list` used for?

Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2628]: Missing `;` too. :)

File: src/librustdoc/html/static/main.js [hunk 2]
```diff
@@ -2621,6 +2624,16 @@ function defocusSearchBar() {
         });
     }());
 
+    function showImportantTraits(content) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2627]: Actually, this function is never used. Please remove it.

File: src/librustdoc/html/static/main.js [hunk 3]
```diff
@@ -2621,6 +2624,16 @@ function defocusSearchBar() {
...
+        let list = content.classList
+    }
+
+    onEachLazy(document.getElementsByClassName("important-traits"), function(e) {
+        e.onclick = function() {
+            e.getElementsByClassName('important-traits-tooltiptext')[0].classList.toggle("force-tooltip")
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2633]: I think it's better to use either `this` or `event.target` in here (`event` would be the first argument given to onclick).

Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2633]: Also missing `;`.

**Comment by GuillaumeGomez** [comment]
You'll need to add the CSS changes inside the newly added `ayu` theme too.

File: src/librustdoc/html/static/main.js [hunk 4]
```diff
@@ -363,6 +363,7 @@ function defocusSearchBar() {
     function handleEscape(ev) {
         var help = getHelpElement();
         var search = getSearchElement();
+        hideModal();
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:366]: Why did you add the `hideModal` calls BTW?

File: src/librustdoc/html/static/themes/dark.css [hunk 5]
```diff
@@ -337,6 +337,11 @@ pre.ignore:hover, .information:hover + pre.ignore {
 	border-color: transparent black transparent transparent;
 }
 
+.important-traits-tooltiptext {
+	background-color: #111 !important;
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/themes/dark.css:341]: Do you really need `!important`? Generally, just giving a more precise selector allows to override things without needing them.

File: src/librustdoc/html/static/main.js (hunk 4 above)
Comment (Manishearth) [reply src/librustdoc/html/static/main.js:366]: Didn't intend to. Will fix.

File: src/librustdoc/html/static/themes/dark.css (hunk 5 above)
Comment (Manishearth) [reply src/librustdoc/html/static/themes/dark.css:341]: Yes, because we're just overriding the same declaration in the themeless version. I added a decl in the themeless version with the intent of having okayish defaults.

File: src/librustdoc/html/static/main.js (hunk 3 above)
Comment (Manishearth) [reply src/librustdoc/html/static/main.js:2633]: Wouldn't that just be `e` anyway?

**Comment by Manishearth** [comment]
Everything addressed except for the Ayu thing. I'll add that shortly after trying out my rebased build

File: src/librustdoc/html/static/themes/dark.css (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/themes/dark.css:341]: Maybe remove it from the themeless version then? We always have themes after all...

**Comment by Manishearth** [comment]
Hmm, i've updated my branch but the PR seems frozen https://github.com/Manishearth/rust/tree/re-spotlight

File: src/librustdoc/html/static/main.js (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:2633]: I'm not too confident about the variable capture process in JS so I prefer to minimize it as much as possible.

**Comment by Manishearth** [comment]
Reopened as https://github.com/rust-lang/rust/pull/74370
